lib/vrl/stdlib/src/encode_logfmt.rs

```diff
@@ -0,0 +1,325 @@
+use std::result::Result;
+use vrl::prelude::*;
+
+mod logfmt {
+    use std::collections::{BTreeMap, HashSet};
+    use std::fmt::Write;
+
+    use shared::btreemap;
+    use Value::{Array, Object};
+
+    use vrl::prelude::*;
+
+    fn encode_string(output: &mut String, str: &str) {
+        let needs_quotting = str.find(' ').is_some();
+        if needs_quotting {
+            output.write_char('"').unwrap();
+        }
+
+        for c in str.chars() {
+            let needs_escaping = matches!(c, '\\' | '"');
+            if needs_escaping {
+                output.write_char('\\').expect("");
+            }
+
+            output.write_char(c).unwrap();
+        }
+
+        if needs_quotting {
+            output.write_char('"').unwrap();
+        }
+    }
+
+    fn encode_value(output: &mut String, value: &Value) {
+        match value {
+            Value::Bytes(b) => {
+                let val = String::from_utf8_lossy(b);
+                encode_string(output, &val)
+            }
+            _ => {
+                let val = format!("{}", value);
+                encode_string(output, &val)
+            }
+        }
+    }
+
+    fn flatten(
+        input: &BTreeMap<String, Value>,
+        parent_key: &str,
+        separator: char,
+        depth: usize,
+    ) -> BTreeMap<String, Value> {
+        let mut items = BTreeMap::<String, Value>::new();
+
+        for (key, value) in input.iter() {
+            let new_key = if depth > 0 {
+                format!("{}{}{}", parent_key, separator, key)
+            } else {
+                key.into()
+            };
+
+            match value {
+                Object(map) => {
+                    let mut res = flatten(map, &new_key, separator, depth + 1);
+                    items.append(&mut res);
+                }
+                Array(array) => {
+                    for (idx, v) in array.iter().enumerate() {
+                        let array_map = btreemap! {
+                            idx.to_string() => v.clone()
+                        };
+                        let mut res = flatten(&array_map, &new_key, separator, depth + 1);
+                        items.append(&mut res);
+                    }
+                }
+                _ => {
+                    items.insert(new_key, value.clone());
+                }
+            };
+        }
+
+        items
+    }
+
+    fn encode_field(output: &mut String, key: &str, value: &Value) {
+        encode_string(output, key);
+        output.write_char('=').unwrap();
+        encode_value(output, value)
+    }
+
+    pub fn encode(input: &BTreeMap<String, Value>, fields: &[String]) -> String {
```

Comment:
I think you could avoid some cloning below if you had `input` just be a `BTreeMap<String, Value>`. We don't need it outside of calling this function.